apiserver/application/application.go

```diff
@@ -1049,54 +1034,107 @@ func (api *API) differentControllerOfferedApplication(userName, modelName, offer
 	//return &offers[0], sourceModelTag, nil
 }
 
-func (api *API) makeOfferParamsFromApplication(offerName string, app *state.Application) (*params.ApplicationOffer, error) {
-	ch, _, err := app.Charm()
+func (api *API) makeOfferParams(st *state.State, offer *jujucrossmodel.ApplicationOffer) (
+	*params.ApplicationOffer, error,
+) {
+	app, err := st.Application(offer.ApplicationName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	appBindings, err := app.EndpointBindings()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	result := params.ApplicationOffer{
-		SourceModelTag:         api.backend.ModelTag().String(),
-		OfferName:              offerName,
-		ApplicationDescription: ch.Meta().Description,
+		OfferName:              offer.OfferName,
+		ApplicationDescription: offer.ApplicationDescription,
 	}
-	eps, err := app.Endpoints()
-	for _, ep := range eps {
+
+	spaceNames := set.NewStrings()
+	for _, ep := range offer.Endpoints {
 		result.Endpoints = append(result.Endpoints, params.RemoteEndpoint{
 			Name:      ep.Name,
 			Interface: ep.Interface,
 			Role:      ep.Role,
 			Scope:     ep.Scope,
 			Limit:     ep.Limit,
 		})
+		spaceNames.Add(appBindings[ep.Name])
+	}
+
+	spaces, err := api.collectRemoteSpaces(st, spaceNames.SortedValues())
+	if errors.IsNotSupported(err) {
+		// Provider doesn't support ProviderSpaceInfo; continue
+		// without any space information, we shouldn't short-circuit
+		// cross-model connections.
+		return &result, nil
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Ensure bindings only contains entries for which we have spaces.
+	result.Bindings = make(map[string]string)
+	for epName, spaceName := range appBindings {
+		space, ok := spaces[spaceName]
+		if !ok {
+			continue
+		}
+		result.Bindings[epName] = spaceName
+		result.Spaces = append(result.Spaces, space)
 	}
 	return &result, nil
 }
 
-func (api *API) sameControllerOfferedApplication(sourceModelTag names.ModelTag, offerName string, perm permission.Access) (
-	_ *state.Application,
-	releaser func(),
-	err error,
-) {
-	defer func() {
-		if err != nil && releaser != nil {
-			releaser()
-		}
-	}()
+func (api *API) collectRemoteSpaces(st *state.State, spaceNames []string) (map[string]params.RemoteSpace, error) {
+	env, err := api.getEnviron(st)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	netEnv, ok := environs.SupportsNetworking(env)
+	if !ok {
+		logger.Debugf("cloud provider doesn't support networking, not getting space info")
+		return nil, nil
+	}
 
-	fail := func(err error) (
-		*state.Application,
-		func(),
-		error,
-	) {
-		return nil, releaser, err
+	results := make(map[string]params.RemoteSpace)
+	for _, name := range spaceNames {
+		space := environs.DefaultSpaceInfo
+		if name != "" {
```

Comment:
I had a `DefaultProviderSpaceId`, but I replaced it with `DefaultSpaceInfo` when I changed the signature of `Environ.ProviderSpaceInfo`. I'll add another for `DefaultSpaceName`.